### DISCUSSION TITLE
Support for mapping frozen dataclasses

### DISCUSSION THREAD

**Discussion by christoph-blessing** [sqlalchemy/sqlalchemy#9192] [Ideas]
The `dataclasses` module in Python allows developers to save time when creating
classes by auto-generating boilerplate code for certain methods. This is done
by decorating a class with the `@dataclass` decorator and providing some
annotations in the class body. 

The mapping of most of these decorated classes (a.k.a dataclasses) is currently
supported by SQLAlchemy. That said dataclasses which are frozen can not be
mapped at the moment.

One can freeze a dataclass by decorating it with `@dataclass(frozen=True)`. The
effect of this is that trying to set an attribute on a frozen dataclass will
raise an exception. Therefore a frozen dataclass is immutable which makes them
a great fit for implementing [value objects] in Python.

Another benefit of freezing a dataclass is that it will be hashable by default
which makes it possible to use it in sets and as keys in dictionaries.

As mentioned before frozen dataclasses can not be mapped by SQLAlchemy at the
moment. This makes it much harder (impossible?) to implement value objects in
Python and map them to database tables. 

It is still possible to map a hashable dataclass by decorating it with
`@dataclass(unsafe_hash=True)` but this makes their use in sets and
dictionaries unsafe because it is possible to modify the objects after they
were added to one of these data structures.

Therefore I think it would be helpful to support the mapping of frozen
dataclasses. Thoughts?

[value objects]: https://www.martinfowler.com/bliki/ValueObject.html

**Comment by zzzeek** [comment]
Hi -

frozen will be a non-trivial implementation and testing effort since nothing within SQLAlchemy's ORM has any notion of fields that are not "writeable", and mutation of fields is actually a given in the mechanics of the ORM.  Such as, if I did `session.refresh(obj)` and the database values changed, now the value is different.  Was that "frozen" or not?     So "frozen" here would at most be a prevention of being able to set values in Python, e.g. `myobject.attr = "new value"`.

Hashability is allowed by using the safe_unhash option, so hashable value objects are certainly doable, it's just that within the ORM, these value objects are not Python-only, they are proxies to a database row which is "live" and can change at any time.   

> As mentioned before frozen dataclasses can not be mapped by SQLAlchemy at the moment. This makes it much harder (impossible?) to implement value objects in Python and map them to database tables.
> 
> It is still possible to map a hashable dataclass by decorating it with `@dataclass(unsafe_hash=True)` but this makes their use in sets and dictionaries unsafe because it is possible to modify the objects after they were added to one of these data structures.

it will always be possible for the objects to be modified, since one can change their value in the database via other means and then refresh.  so the question of "possible" needs to be graded on a scale here.  You can modify the values in a real frozen dataclass also by manipulating `__dict__`.    so in Python, "frozen" is more of a convention that's followed.    I dont see any of this as "much harder" and more like "less formally correct".

> Therefore I think it would be helpful to support the mapping of frozen dataclasses. Thoughts?

it's definitely valid for the road map but not anytime soon, we have just the initial 2.0 release to get working first as the many new features all have more immediate improvements to be made.

**Comment by christoph-blessing** [comment]
Thanks for reading my post and giving feedback.

I understand that frozen dataclasses are not truly immutable and I don't expect SQLAlchemy to enforce immutability somehow. I would already be happy if a mapped frozen dataclass would prevent my own code from modifying it using the normal mechanism for setting attriubtes.

**Reply by zzzeek** [reply]
right that's what we can do here, is add an event that will just raise for ordinary setattr operations.   it still will be possible for things like lazy loading to populate unloaded attributes, refreshes to change them, etc.      a program that is relying very heavily on hashability of ORM objects however could have subtle bugs if objects have values changing in the database.    

as stated before, the frozen feature, and actually much moreso the "slots" feature, are things we would add a little later.  we are still trying to get the basic dataclass feature to work as expected with the current featureset and we are not there yet.

**Reply by christoph-blessing** [reply]
Okay makes sense. Thank you.

**Comment by CaselIT** [comment]
I've created an use case issue to better track this feature https://github.com/sqlalchemy/sqlalchemy/issues/9197

**Reply by christoph-blessing** [reply]
Thank you. I wasn't sure if a use case issue is correct here because its description did not sound like a good fit for this feature.

**Reply by CaselIT** [reply]
no problem, it's fine to use a discussion to gauge the interest for new feature and also discuss implementation details.
